### ISSUE TITLE
SharedArrayBuffer + Atomics

### ISSUE THREAD

**Issue by flagxor** [w3ctag/design-reviews#165]
Hello TAG!

I'm requesting a TAG review of:

  - Name: SharedArrayBuffer

  - Specification URL:
     https://tc39.github.io/ecma262/
     (Sections referencing Atomics / SharedArrayBuffer)
     Last separate draft:
    https://tc39.github.io/ecmascript_sharedmem/shmem.html

  - Explainer, Requirements Doc, or Example code:
Docs:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/SharedArrayBuffer
Some demo links:
https://flagxor.github.io/sab-demo/convolve-multicore/convolve.html
https://browsix.org/latex-demo-sync/
(You must enable under about:flags in Chrome)

  - Primary contacts: flagxor, binji

Further details (optional):
  - Relevant time constraints or deadlines:
Seek further clarification on any considerations as the feature is incorporated into the web platform.

  - [x] I have read and filled out the [Self-Review Questionnare on Security and Privacy](https://www.w3.org/TR/security-privacy-questionnaire/).
Here are the answers:
https://github.com/flagxor/sab-tag-questions

  - [x] I have reviewed the TAG's [API Design Principles](https://w3ctag.github.io/design-principles/)

You should also know that...

We probably should have reached out sooner, as this feature has web platform impact but has been merged with ECMAscript.

We'd prefer the TAG provide feedback as (please select one):

  - [x] open issues in our Github repo for each point of feedback
https://github.com/tc39/ecmascript_sharedmem

  - [ ] open a single issue in our Github repo for the entire review
  - [ ] leave review feedback as a comment in this issue and @-notify [github usernames]

**Comment by cynthia** [comment]
Also related: https://tc39.github.io/ecma262/#sec-memory-model

**Comment by dbaron** [comment]
We just spent a bit of time discussing this in a breakout at the TAG face-to-face meeting.

Given that this is a space that the relevant TC39 folks have a lot more expertise in than the TAG does, it's not clear how much we're going to be able to add.

Our current assumption is that this is mainly going to be used by (a) JavaScript or WebAssembly as a compilation target for code written in other languages or (b) low-level libraries written by a very small number of people.  Code dealing with mutable shared state is very difficult to write, which is why the Web's model for multithreaded code emphasizes workers, and we expect that workers will continue to be the main emphasis.

We had a brief discussion about `SharedArrayBuffer` not being transferable, and agree that aspect of the design makes sense.  We note that this means that some code that works with `ArrayBuffer` no longer works with `SharedArrayBuffer`, but that's probably a good thing since converting code from copying to sharing requires very careful consideration.

One of the interesting pieces is that this introduces the possibility of deadlocks into the Web platform.  We're curious what the plan for handling deadlocks is.  Is there going to be some deadlock detection mechanism, or is it the plan simply to treat deadlocks like a tab crash, or is that something that's going to be left more open to implementations?

Thanks for bringing this up for TAG review.  I think we're nearly ready to close this issue, although we'd be interested to hear feedback on deadlocks.

**Comment by travisleithead** [comment]
We really, really, really needed a good explainer for this. MDN didn't cut it for me.

**Comment by annevk** [comment]
> We had a brief discussion about SharedArrayBuffer not being transferable, and agree that aspect of the design makes sense. We note that this means that some code that works with ArrayBuffer no longer works with SharedArrayBuffer, but that's probably a good thing since converting code from copying to sharing requires very careful consideration.

To be clear, if you just copy ArrayBuffer now (not transfer), SharedArrayBuffer can easily replace that (except you end up with shared memory, rather than copied memory). If you transfer an ArrayBuffer the original ends up detached, that isn't possible because SharedArrayBuffer doesn't do detachment.

**Comment by dbaron** [comment]
@flagxor & @binji: curious if you have any feedback on this piece of https://github.com/w3ctag/design-reviews/issues/165#issuecomment-298149776 (or know someone else we should ask):
> One of the interesting pieces is that this introduces the possibility of deadlocks into the Web platform. We're curious what the plan for handling deadlocks is. Is there going to be some deadlock detection mechanism, or is it the plan simply to treat deadlocks like a tab crash, or is that something that's going to be left more open to implementations?

**Comment by binji** [comment]
@lars-t-hansen (hm, this doesn't seem to be working).

Sorry, I didn't notice this thread until now.

Deadlocks are possible to write, but there are provisions in the spec to disallow calls to `Atomics.wait` in certain contexts, the only call that can block. Currently, `Atomics.wait` is only allowed in DedicatedWorkers and SharedWorkers (see https://github.com/whatwg/html/pull/2521). As a result, this shouldn't hang the tab, and manifest as just a bug in the page (e.g. the page is waiting for the worker to complete, so perhaps some operation is pending, but the page will not be hung). The result would be similar to writing `while (true) {}` in a worker, with the difference being that `Atomics.wait` is not actually a spin loop.

As for detecting deadlocks, I'm not certain any of the implementations do this, or if it is even possible given the low-level nature of the wake/wait mechanism. You could easily determine that there are two workers are waiting via `Atomics.wait`, but I don't know if you could tell that they were related.

**Comment by dbaron** [comment]
Thanks.  `[[CanBlock]]` being false for main thread JS sounds to me like a decent compromise.

**Comment by lars-t-hansen** [comment]
Indeed I think that the only situation in which we *must* have `[[CanBlock]]` equal to `true` is in dedicated workers.  The recent modifications of the HTML spec to accomodate this feature, as @binji notes, also allows it in shared workers (not yet implemented in Firefox), but not in service workers, nor in the main thread nor elsewhere (if there is any elsewhere).

We discussed at one point whether the `Worker` constructor should even take a flag specifying the value of this flag, mainly to force the browser to allocate a dedicated thread to the worker if the flag was `true` (as opposed to sharing a thread among workers or delaying the start of a worker, as Firefox did (does?)), but this fell by the wayside as the worker spec was updated and we decided browsers could always afford dedicated threads for their workers.

I think that it is possible in principle to have automatic, no-false-positives deadlock detection but that it will be almost worthless in practice given that nearly always, some agent in the cluster will be nonblocking and have access to the shared memory, and can therefore in principle always be available to wake some waiting agent.  It is only in the situation where all the nonblocking agents have dropped their last pointers to any shared memory and have run GC to prove that, that we can potentially detect deadlocks among the remaining agents.

The spec arguably provides too little in terms of introspection, though the mechanisms that are provided are so low level, and shared memory is so powerful for inter-agent communication, that it's not hard to imagine building the necessary introspection and deadlock detection facilities in JS.  For example, an agent can account for itself in shared memory before going into a wait, thereby allowing the main thread or a dedicated aux thread to detect whether its agents are deadlocked.  For the same reason, we have no interrupt mechanism - the application is expected to provide one with the appropriate semantics.

I also think that as time goes by we'll see the need for specific abstractions that make it simpler and less error prone to use shared memory, and we'll push those abstractions into specs.  Along those lines, I have one proposal pending here: https://github.com/lars-t-hansen/moz-sandbox/blob/master/sab/waitNonblocking.md
